### PR TITLE
Restyle the display of media elements and archive items

### LINKED ISSUE #219: Restyle the media elements
We'll be restyling the media elements to be consistent with our interface but also reflective of their original publishing platform and format.

**Notes:**
- These screenshots are quick mockups in Figma and don't reflect final colors or spacing.
- Where the platform supports a handle/username distinct from a profile name, we include it next to the profile name. (AFAIK, only YouTube doesn't consistently support this.)
- Terminology like "followers" or "subscribers" will reflect the platform.
- The positioning of the caption or text associated with a piece of media will be platform-specific. E.g., YouTube and Instagram put the caption below the media, so we respect that. Facebook and Twitter put it above, so we do the same.
- We will truncate long captions behind a "Read more" link. (However, I'm going to make it smart enough not to truncate if there's only some small number of additional characters remaining.)
- The platform is currently noted twice: in the follower/subscriber count and in the Published timestamp. The icon is used in the latter.
- The layout will be responsive; e.g., on smaller screens, the profile image will be aligned with the rest of the content rather than hanging off the edge.
- We will support RTL layout from the beginning.
- Single videos/images will be presented in original aspect ratio. Image/video sets will be cropped to square and arranged on a grid. The grid will adapt based on number of media in the set, e.g. 1x3 or 2x2 etc. Clicking images or videos in a set will bring up the lightbox.
- Single videos will be playable inline. Videos in sets will only be playable from the lightbox.
- Click targets on the media will be figured out during development. Currently, I'm sure the author header will link to the author detail page; the Archived timestamp will link to our show page for this media resource; the Published timestamp will link to the original URL; media will link as mentioned above (either playable or lightbox). The question is whether the body of the media — comment/caption — should link to the media show page, too. (I think not, especially since there are so many other possible interactions in that content, like clickable inline links.)

**Twitter:**
image

**Instagram:**
image

**YouTube:**
image

**Facebook:**
image

### LINKED ISSUE #222: Detect and auto-link URLs in archive captions
Captions on archived items often contain URLs. It would be nice if they could be auto-linked.

1. **Discussion:** Is this a bad idea?
2. **Implementation:** [This SO thread](https://stackoverflow.com/questions/3665072/how-to-extract-urls-from-text) mentions some simple ways (e.g., `URI.extract`) but implies it fails on some common edge cases, so [the `postrank-uri` gem](https://github.com/postrank-labs/postrank-uri) is suggested as safer.

### LINKED ISSUE #226: Truncate long media captions
Currently (as of #220), we don't truncate the text/caption associated with a piece of media. For most Twitter and YouTube, this is fine, since those are naturally capped at a sensible length. But Facebook and Instagram can have _long_ captions, which gunk up the page.

As part of the original design of #219, we wanted to truncate past a certain length with a "Read more" link. I would have preferred to do this entirely in CSS, but it will necessitate JS.

### PR DESCRIPTION
**Pull request by reefdog**
This PR delivers the bulk of #219, restyling the display of archive items on the index page.

**Before:**
Screenshot of an archive item, before redesign

**After:**
Screenshot of an archive item, after redesign

**Notes:**
- These features were determined to be just out of scope, and will be done as separate work: #225.
- So far, this styling only applies on the archive index page. Applying it to other pages has been bumped to #224.

**Testing:**
- `yarn install` to get a new package (Day.js)
- `./bin/dev`
- Log in and go to the root of the site ("Most Recent Archived Media") and observe the new item design. Especially:
   - Scrape some multi-image tweets or Instagram posts and make sure they display as expected
   - Click author headers to go to author pages
   - Pay attention to "Published" and "Archived" timestamps, which should appear in your local time

**Feedback wanted:** Obviously, any problems you find, but also **your thoughts/criticisms of the design** and **any concerns you have with implementation**. In particular, I made heavy use of the `ArchivableItem` module, and even (😬) exposed the Rails URL helpers to the models (see `author_canonical_path` in `normalized_attrs_for_views` in any source model). Please let me know if you're uncomfortable with any of this!

Closes #219 
Closes #222
Closes #226

### CODE REVIEW

**Review by oneroyalace — COMMENTED** [review]
Most of what I see LGTM, including the model-specific view attributes! I think there's a tiny bug that occurs after adding media, though. 

Immediately after archiving a URL, I see the old archive view without your view/style updates. Post order for the archive view also switches to chronological (intsead of reverse chronological). 

Screen Shot 2022-06-03 at 3 54 05 PM


When I refresh the page, everything looks as expected. 

Screen Shot 2022-06-03 at 3 54 14 PM

### Steps to reproduce the bug: 
1. Manually archive https://twitter.com/TheStalwart/status/1532704755847467008
2. You'll see an unexpected view
3. Refresh
4. You'll see the espected view

I suppose this may not matter, since media will be added via factstream.

**Comment by reefdog** [comment]
@oneroyalace Ah yes! I'd noticed that and convinced myself it could wait for #224, especially with archiving a URL being a dev-only feature, but I think for testing purposes alone it makes sense to resolve it here. Thanks for the note!

File: app/helpers/archive_helper.rb [hunk 1]
```diff
@@ -1,3 +1,12 @@
 # typed: strict
 module ArchiveHelper
+  def humanized_community_count(number, noun)
+    count = number_to_human(number, units: {
```
[the remark this answers is not in the document]
Comment (reefdog) [reply app/helpers/archive_helper.rb:4]: Yeah, it's a neat one. Although lol at my including a unit for trillions of followers. That's… optimistic.

File: app/models/archivable_item.rb [hunk 2]
```diff
@@ -2,7 +2,52 @@
...
+    self.try(:images).try(:empty?) == false || self.try(:videos).try(:empty?) == false
+  end
+
+  # Combines images and videos into a single array.
+  #
+  # This is because an archivable item may be a mixed-image/video collection and we need to be able
+  # to represent it.
+  #
+  # NB: Currently this sorts by creation date. This is a coarse and inaccurate proxy for the actual
+  #     original display order, and a future improvement store and utilize actual display order.
```
[the remark this answers is not in the document]
Comment (reefdog) [reply app/models/archivable_item.rb]: That's a good question that I hadn't considered. I assumed so… or, if they didn't make it explicit, perhaps there was a more accurate implicit order (like array order or logical display order) that we could adopt as the closest thing to an explicit order. Any of that is better than my using our own `created_at`!

**Comment by reefdog** [comment]
@oneroyalace I've resolved that issue, so all should be well now. Mind giving it another 👀? (Offer also available to @cguess if Asa is off for the night.)

**Pull request merged**